flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/io/BufferBlockerTest.java

```diff
@@ -0,0 +1,322 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.streaming.runtime.io;
+
+import org.apache.flink.core.memory.MemorySegment;
+import org.apache.flink.core.memory.MemorySegmentFactory;
+import org.apache.flink.runtime.io.network.buffer.Buffer;
+import org.apache.flink.runtime.io.network.buffer.FreeingBufferRecycler;
+import org.apache.flink.runtime.io.network.buffer.NetworkBuffer;
+import org.apache.flink.runtime.io.network.partition.consumer.BufferOrEvent;
+import org.apache.flink.streaming.runtime.io.CreditBasedBufferBlocker.BufferOrEventSequence;
+
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.Random;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertNull;
+import static org.junit.Assert.assertTrue;
+import static org.junit.Assert.fail;
+
+/**
+ * Tests for {@link CreditBasedBufferBlocker}.
+ */
+public class BufferBlockerTest {
+
+	private static final int PAGE_SIZE = 4096;
+
+	private CreditBasedBufferBlocker bufferBlocker;
+
+
+	// ------------------------------------------------------------------------
+	//  Setup / Cleanup
+	// ------------------------------------------------------------------------
+
+	@Before
+	public void createSpiller() {
+		bufferBlocker = new CreditBasedBufferBlocker(PAGE_SIZE);
+	}
+
+	@After
+	public void cleanupSpiller() {
+		if (bufferBlocker != null) {
+			bufferBlocker.close();
+		}
+	}
+
+	// ------------------------------------------------------------------------
+	//  Tests
+	// ------------------------------------------------------------------------
+
+	@Test
+	public void testRollOverEmptySequences() {
+		assertNull(bufferBlocker.rollOver());
+		assertNull(bufferBlocker.rollOver());
+		assertNull(bufferBlocker.rollOver());
+	}
+
+	@Test
+	public void testSpillAndRollOverSimple() {
+		final Random rnd = new Random();
+		final Random bufferRnd = new Random();
+
+		final int maxNumEventsAndBuffers = 3000;
+		final int maxNumChannels = 1656;
+
+		// do multiple blocking / rolling over rounds
+		for (int round = 0; round < 5; round++) {
```

Comment:
Can you deduplicate the code of those two unit tests (`testSpillAndRollOverSimple` and `testSpillWhileReading`)? It seems like this one is just a one sequence of the next one?